### ISSUE TITLE
4.8-Classification

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#35]
### 把Classification當成Regression

Regression-輸入一個向量(vector)x，輸出一個數值y，並期望此數值與正解y hat相近

想法一:把目標y hat Class編號，變成數值(class 1/class 2/...)→不是一個好方法

![image](https://user-images.githubusercontent.com/34474924/236364879-d4577171-f5ca-4a4b-850e-ec57e325c78a.png)

觀察Model輸出的數值y，與目標Class的編號的關係
問題：Class的編號表示會預設其之間的關係
e.g. Class 1與Class 2 關係較近，Class 1與Class 3關係較遠→不是一個好方法

# Class as one-hot vector
方法二:把Class 用one-hot vector表示→比較常見

![image](https://user-images.githubusercontent.com/34474924/236365988-7245ffca-622c-40e3-bf42-29e590cbe139.png)

N個Class用N個 N * 1 維陣列表示（合併就是N維的單位矩陣）

### 把output改成3個數值
目標y hat Class是有3個數值的向量，output應該也要有3個數值
怎麼把network從1個數值改成輸出3個數值？→把output方法重複3次：
a1 a2 a3 乘上不同的weight，加上不同的bias來輸出不同的y1 y2 y3

![image](https://user-images.githubusercontent.com/34474924/236458039-642d110e-1723-4652-8e37-53032dd68119.png)

Claassification作法小結
Regression是input x ，output y，y與y hat越接近越好
Claassification也是input x，output y，但y是一個向量，y與y hat越接近越好
但做Claassification時，會把y再通過Function `Softmax` 得出y'當作最終輸出

### Softmax

Softmax會把值介於0~1之間, 容易與y hat計算相似度，

![image](https://user-images.githubusercontent.com/34474924/236590388-b3429f01-44a8-477b-a1b0-84737ba6c100.png)

Softmax怎麼運作？
1.將y取以e為底數的指數函數，這樣y的值就會變成正的
2.除以所有步驟1的和，得出y'
所有output參數做完步驟1與2後，範圍會變成 1>y'>0， 其所有參數和為1

![image](https://user-images.githubusercontent.com/34474924/236459420-e02b805a-9634-4edf-824a-34378f38efc6.png)

e.g. 今天要做一個關於動物的分類，分類結果有鳥類、哺乳類、魚類，
將分類結果編號並做one-hot vector→有3個結果，做成3個 3 * 1 維陣列
鳥類	[1 0 0] 
哺乳類[0 1 0]		
魚類	[0 0 1]
(合併就是3維的單位矩陣)
假設output y1 y2 y3 = [3 1 -3]
通過Softmax→y1' y2' y3' = [0.88 0.12 0]
可發現與鳥類最接近

Softmax功能
1.Normalize，讓所有y'範圍為0到1之間，且和為1
2.加深值之間的差距

Softmax的輸入通常較logit

___________

logit 意思
![image](https://user-images.githubusercontent.com/34474924/236465088-246a90c4-fa30-44bf-ae35-a5bb54dcd34a.png)

![image](https://user-images.githubusercontent.com/34474924/236465110-bd1d5b5a-9e55-41ea-9263-84fecd432e93.png)

從概率P->Odds->Logit
這就是一個Logit變換。實際上，所謂 Logit 模型可以理解成 Log-it（即 it 的自然對數——這里的 it 指的就是Odds）

與概率不同，Logit的一個很重要的特性就是沒有上下限——這就給建模帶來極大方便

_______________

計算距離e有不同的做法：Mean Square Error與Cross-entropy
![image](https://user-images.githubusercontent.com/34474924/236590425-ad84f6ee-eb6b-438d-89b3-bb11eded19ec.png)

P.S. pytorch裡面Cross-entropy會內建Softmax

現在要做一個3個Class的分類，input x，output y1 y2 y3，把y們丟到Softmax得y'們
正解是[1 0 0]，要計算y'們與y hat的距離e
將y3設為不影響結果的數值，只觀察y1 y2變動的結果
e設為Mean Square Error與Cross-entropy的不同
當y1大y2小→接近正解[1 0 0]，Loss小
當y1小y2大→Loss大

![image](https://user-images.githubusercontent.com/34474924/236590622-5d864604-f7d5-446f-86ac-fd0fb91dd58e.png)

可發現當Loss很大的時候
Cross-entropy有斜率
Mean Square Error比較平坦→可能會卡住，雖然可以用Learning rate解決，但Traning起步仍是較困難的
